### PR TITLE
Update numba to 0.56.3

### PR DESCRIPTION
**Pull request by AndrewVallette**
## Changes

- Updated numerous version requirements: `llvmlite`, `numpy`, CUDA, etc
- Added `ipykernel`, `pygments`, and `pexpect` for optional tests
- Removed pin_compatible for `llvmlite` and `numpy` in favor of more explicit pinnings
- Updated tests (disable AVX512_SKX if detected with `numpy >=1.22` due to low accuracy)
- Updated `setuptools` upper limit
- Removed outdated patch

## Links

- GitHub: https://github.com/numba/numba/tree/0.56.3
- numba recipe: https://github.com/numba/numba/blob/0.56.3/buildscripts/condarecipe.local/meta.yaml
- conda-forge feedstock: https://github.com/conda-forge/numba-feedstock

## Notes

- `0.56.2` was skipped due to the `setuptools <60` requirement which could be problematic for some users. This upper limit was removed in `0.56.3`.
- `setuptools` remains a `run` dependency for now. For in-depth discussions see https://github.com/numba/numba/pull/8366 and https://github.com/numba/numba/pull/8475.
- The upper limit for `numpy` is set to `<1.23`, due to a recently-discovered CUDA compilation failure: https://github.com/numba/numba/issues/8529.
- GPU tests with CUDA were performed outside of Prefect, and passed for `linux-64` and `win-64`.

### CODE REVIEW

**Review by skupr-anaconda — APPROVED** [review]
LGTM

**Review by varlackc — COMMENTED** [review]
`pip` is mentioned as a requirement
https://github.com/AnacondaRecipes/numba-feedstock/blob/720a60ef2e45579e16c0322161368b123b0bb984/recipe/meta.yaml#L37

We probably might need to add a `pip check` to the test section.

**Comment by AndrewVallette** [comment]
> `pip` is mentioned as a requirement
> 
> https://github.com/AnacondaRecipes/numba-feedstock/blob/720a60ef2e45579e16c0322161368b123b0bb984/recipe/meta.yaml#L37
> 
> We probably might need to add a `pip check` to the test section.

`pip check` is performed in the test scripts. The `numba` team and `conda-forge` are doing it like this, so I opted to do the same.

**Review by ryanskeith — APPROVED** [review]
LGTM

File: recipe/meta.yaml [hunk 1]
```diff
@@ -34,35 +31,42 @@ requirements:
...
+    # NumPy 1.22.0, 1.22.1, 1.22.2 are all broken for ufuncs
+    # xref: https://github.com/numba/numba/issues/7756
+    # NumPy >=1.23 breaks with CUDA test: test_reinterpret_array_type
+    # xref: https://github.com/numba/numba/issues/8529
+    - numpy >=1.18,!=1.22.0,!=1.22.1,!=1.22.2,<1.23
     # needed for pkg_resources
     - setuptools
 
   run_constrained:
     - tbb  >=2021.0
```
Comment (cbouss) [inline recipe/meta.yaml]: Should tbb upper bound be aligned with the host definition? tbb-devel 2021.*,<2021.6

File: recipe/meta.yaml [hunk 2]
```diff
@@ -34,35 +31,42 @@ requirements:
...
-    - patch     # [not win]
-    - m2-patch  # [win]
+
   host:
     - python
     - pip
     - setuptools
     - wheel
-    - llvmlite 0.38.*
+    - llvmlite 0.39.*
```
Comment (cbouss) [inline recipe/meta.yaml:48]: Are pip and wheel required? I am not sure, but it likely doesn't hurt to leave them here.

File: recipe/meta.yaml [hunk 3]
```diff
@@ -34,35 +31,42 @@ requirements:
...
+    - cudatoolkit >=10.2
     # scipy 1.0 or later
     - scipy >=1.0
-    # numpy has a hard upper limit currently
-    - numpy >=1.18,<1.22
+    # CUDA Python 11.6 or later
+    - cuda-python >=11.6
 
 test:
   requires:
```
Comment (cbouss) [inline recipe/meta.yaml:80]: ipykernel is in numba's recipe:
https://github.com/numba/numba/blob/0.56.3/buildscripts/condarecipe.local/meta.yaml#L85

pygments and pexpect could also be added:
https://github.com/numba/numba/blob/0.56.3/buildscripts/incremental/setup_conda_environment.sh#L62

tbb also has an upper bound:
https://github.com/numba/numba/blob/0.56.3/buildscripts/incremental/setup_conda_environment.sh#L107

File: recipe/meta.yaml [hunk 4]
```diff
@@ -34,35 +31,42 @@ requirements:
...
+    - tbb-devel 2021.*,<2021.6
+    - _openmp_mutex            # [linux]
+    - importlib_metadata       # [py<39]
+
   run:
     - python
-    - {{ pin_compatible('llvmlite', max_pin='x.x') }}
-    - {{ pin_compatible('numpy') }}
+    - importlib_metadata       # [py<39]
+    - llvmlite >=0.39.*,<1
```
Comment (cbouss) [inline recipe/meta.yaml]: should be llvmlite >=0.39.*,<0.40
https://github.com/numba/numba/blob/0.56.3/setup.py#L28

File: recipe/meta.yaml [hunk 5]
```diff
@@ -34,35 +31,42 @@ requirements:
...
-    - m2-patch  # [win]
+
   host:
     - python
     - pip
     - setuptools
     - wheel
-    - llvmlite 0.38.*
+    - llvmlite 0.39.*
     - numpy
```
Comment (cbouss) [inline recipe/meta.yaml]: I think numpy should have the same restrictions in host and run:
numpy >=1.18,!=1.22.0,!=1.22.1,!=1.22.2,<1.23
See: https://github.com/numba/numba/blob/0.56.3/setup.py#L373

**Review by cbouss — CHANGES_REQUESTED** [review]
Mostly, I think the upper bounds should be updated in the run, run_constraint and test/requires sections for tbb and llvmlite

File: recipe/meta.yaml (hunk 4 above)
Comment (AndrewVallette) [reply recipe/meta.yaml]: Thats right, thanks!

File: recipe/meta.yaml (hunk 1 above)
Comment (AndrewVallette) [reply recipe/meta.yaml]: I don't think `tbb` and `tbb-devel` have the same pinnings here. The documentation covers the requirements of TBB in depth here: https://numba.readthedocs.io/en/0.56.3/user/installing.html

- **tbb-devel** - provides TBB headers/libraries for compiling TBB support into Numba’s threading backend (`2021 <=` version `< 2021.6` required).
- **tbb** - provides the TBB runtime libraries used by Numba’s TBB threading backend (version `>= 2021` required).

**Edit:** Hmm, maybe a `2021.*` or `<2022` would be safer. Which do you think would be better?  I still think the `<2021.6` isn't quite right here -- the documentation conveys that it should work with any 2021.x version.

File: recipe/meta.yaml (hunk 2 above)
Comment (AndrewVallette) [reply recipe/meta.yaml:48]: I'm not sure either, but I agree it's fine as-is.

File: recipe/meta.yaml (hunk 3 above)
Comment (AndrewVallette) [reply recipe/meta.yaml:80]: Sure, `ipykernel`, `pygments`, and `pexpect` can be added, I can see they're used in some optional tests.

My `tbb` comments are here: https://github.com/AnacondaRecipes/numba-feedstock/pull/10#discussion_r1004014248

File: recipe/meta.yaml (hunk 5 above)
Comment (skupr-anaconda) [reply recipe/meta.yaml]: But if in `host` will be `numpy >=1.18,!=1.22.0,!=1.22.1,!=1.22.2,<1.23` then in `run` will be the **latest** numpy version and not the **oldest**.

File: recipe/meta.yaml (hunk 1 above)
Comment (cbouss) [reply recipe/meta.yaml]: I think adding <2022 is the right thing to do. Especially, from what I heard, tbb uses 2021 and 2022 not as the year but as the major version.

Comment (AndrewVallette) [reply recipe/meta.yaml]: Thanks, I thought I remembered they do something a bit odd with the version here.

File: recipe/meta.yaml (hunk 5 above)
Comment (cbouss) [reply recipe/meta.yaml]: Right, because of numpy's run export.
In that case numpy version should be set to the lowest supported: numpy 1.18

**Comment by AndrewVallette** [comment]
> Linter check failed, with the following output: The following problems have been found:
> 
> WARNING: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:0: uses_setuptools: The recipe uses setuptools in run depends ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:45: version_constraints_missing_whitespace: Packages and their version constraints must be space separated ERROR: /tmp/abs_5a4ybtzi5d/clone/recipe/meta.yaml:48: version_constraints_missing_whitespace: Packages and their version constraints must be space separated Errors were found

The linter is incorrect, and I have opened an issue: https://github.com/AnacondaRecipes/numba-feedstock/pull/10

File: recipe/meta.yaml (hunk 5 above)
Comment (AndrewVallette) [reply recipe/meta.yaml]: To get this one working was quite a bit more complicated. I can see Serhii needed to do this before, so he might remember that `numpy` is tricky :)

I am using a `numpy_min_ver` variable to set the appropriate minimum version, depending on platform and Python version.

See the following commits: https://github.com/AnacondaRecipes/numba-feedstock/commit/d3b3890c29fe8fc959f7da3021e3ba94100d59b0 https://github.com/AnacondaRecipes/numba-feedstock/pull/10/commits/80f942b4a22347dcd839b3b54c65480ff310a286

**Comment by AndrewVallette** [comment]
I believe all of the `numpy` pinnings have been resolved, as this is building for all platforms & Python versions now.

The only remaining issue is a DeprecationWarning in `pywin32` for one of the tests for `win-64` on `py310`. The warning is not an issue with `numba` itself.

The relevant part of the test failure is shown here:

```
INFO   FAIL: test_no_accidental_warnings (numba.tests.test_import.TestNumbaImport)
...
INFO     File "C:\b\abs_150hstvvni\croot\numba_1666764588839\_test_env\lib\site-packages\win32\lib\pywintypes.py", line 51, in __import_pywin32_system_module__
INFO       found = _win32sysloader.GetModuleFilename(filename)
INFO   DeprecationWarning: getargs: The 'u' format is deprecated. Use 'U' instead.

```

This seems to occur because our version of `pywin32` (302) is a little out of date. It's my opinion that this should not hold up the release, as it seems to be benign.

**Comment by cbouss** [comment]
I agree. Based on this [link](https://docs.python.org/3/c-api/arg.html#strings-and-buffers) this will be removed in python 3.12.

**Comment by AndrewVallette** [comment]
Re-triggering build due to test `win-64` test timeouts..

**Comment by AndrewVallette** [comment]
Prefect is having issues. Restarting again.

**Comment by skupr-anaconda** [comment]
Only `win64` tests failed:
`AssertionError: Process timed out.`

**Review by cbouss — APPROVED** [review]
LGTM!

**Pull request merged**